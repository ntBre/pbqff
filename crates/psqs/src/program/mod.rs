use std::{
    error::Error, fmt::Display, path::Path, str::FromStr, time::SystemTime,
};

use serde::{Deserialize, Serialize};
use symm::Atom;

use crate::geom::Geom;

pub mod cfour;
pub mod dftbplus;
pub mod molpro;
pub mod mopac;

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ProgramResult {
    pub energy: f64,
    pub cart_geom: Option<Vec<Atom>>,
    pub time: f64,
}

#[derive(Debug, PartialEq, Eq)]
pub enum ProgramError {
    FileNotFound(String),
    ErrorInOutput(String),
    EnergyNotFound(String),
    EnergyParseError(String),
    GeomNotFound(String),
    ReadFileError(String, std::io::ErrorKind),
}

impl ProgramError {
    /// Returns `true` if the program error is [`ErrorInOutput`].
    ///
    /// [`ErrorInOutput`]: ProgramError::ErrorInOutput
    #[must_use]
    pub fn is_error_in_output(&self) -> bool {
        matches!(self, Self::ErrorInOutput(..))
    }
}

impl Display for ProgramError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

impl Error for ProgramError {}

#[derive(Debug, PartialEq, Eq, Copy, Clone)]
pub enum Procedure {
    Opt,
    Freq,
    SinglePt,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Template {
    pub header: String,
}

impl Template {
    pub fn from(s: &str) -> Self {
        Self {
            header: s.to_string(),
        }
    }
}

impl From<String> for Template {
    fn from(header: String) -> Self {
        Self { header }
    }
}

impl FromStr for Template {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            header: s.to_string(),
        })
    }
}

/// A trait for describing programs runnable on a [crate::queue::Queue]
pub trait Program {
    /// returns the file associated with the program's input. it should not
    /// include an extension
    fn filename(&self) -> String;

    /// return the output of `self.filename()` with ".out" appended
    fn outfile(&self) -> String {
        self.filename() + ".out"
    }

    /// return the input file associated with `self`
    fn infile(&self) -> String;

    /// set `filename`
    fn set_filename(&mut self, filename: &str);

    /// the template for writing input files
    fn template(&self) -> &Template;

    /// the file extension for the input file
    fn extension(&self) -> String;

    /// molecular charge
    fn charge(&self) -> isize;

    /// write the input file to the name returned by `filename`
    fn write_input(&mut self, proc: Procedure);

    /// read the output file `filename`
    fn read_output(filename: &str) -> Result<ProgramResult, ProgramError>;

    /// Return all the filenames associated with the Program for deletion when
    /// it finishes
    fn associated_files(&self) -> Vec<String>;

    fn new(
        filename: String,
        template: Template,
        charge: isize,
        geom: Geom,
    ) -> Self;

    /// Build the jobs described by `moles` in memory, but don't write any of
    /// their files yet
    fn build_jobs(
        moles: Vec<Geom>,
        dir: impl AsRef<Path>,
        start_index: usize,
        coeff: f64,
        job_num: usize,
        charge: isize,
        tmpl: Template,
    ) -> Vec<Job<Self>>
    where
        Self: std::marker::Sized,
    {
        let mut count: usize = start_index;
        let mut job_num = job_num;
        let mut jobs = Vec::new();
        for mol in moles {
            let filename = format!("job.{job_num:08}");
            let filename =
                dir.as_ref().join(filename).to_str().unwrap().to_string();
            job_num += 1;
            let mut job = Job::new(
                Self::new(filename, tmpl.clone(), charge, mol.clone()),
                count,
            );
            job.coeff = coeff;
            jobs.push(job);
            count += 1;
        }
        jobs
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Job<P: Program> {
    pub program: P,
    pub pbs_file: String,
    pub job_id: String,

    /// the index in the output array to store the result
    pub index: usize,

    /// the coefficient to multiply by when storing the result
    pub coeff: f64,

    /// the last modified time of `program`'s output file
    pub(crate) modtime: SystemTime,
}

impl<P: Program> Job<P> {
    pub fn new(program: P, index: usize) -> Self {
        Self {
            program,
            pbs_file: String::new(),
            job_id: String::new(),
            index,
            coeff: 1.0,
            modtime: SystemTime::UNIX_EPOCH,
        }
    }

    /// return the current modtime of `self.program`'s output file, or
    /// `self.modtime` if there is an error accessing the metadata
    pub fn modtime(&self) -> SystemTime {
        let p = self.program.outfile();
        if let Ok(meta) = std::fs::metadata(p) {
            meta.modified().unwrap()
        } else {
            self.modtime
        }
    }
}

/// parses the `nth` field of `line` into a float and returns
/// [ProgramError::EnergyParseError] containing `outname` if it fails. a string
/// containing `outname` is allocated in the Err case
#[inline]
fn parse_energy(
    line: &str,
    nth: usize,
    outname: &str,
) -> Result<Option<f64>, ProgramError> {
    line.split_whitespace()
        .nth(nth)
        .map(str::parse::<f64>)
        .transpose()
        .map_err(|_| ProgramError::EnergyParseError(outname.to_owned()))
}
