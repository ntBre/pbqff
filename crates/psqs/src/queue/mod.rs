use std::{
    cell::LazyCell,
    collections::{HashMap, HashSet},
    path::Path,
    process::Command,
    str,
    time::Duration,
};

use crate::{
    geom::Geom,
    program::{Procedure, Program, ProgramError},
};
use crate::{
    program::{Job, ProgramResult},
    time,
};

pub mod local;
pub mod pbs;
pub mod slurm;
use drain::*;
use serde::{Deserialize, Serialize};
mod drain;

pub use drain::Check;

#[derive(PartialEq, Eq, Debug)]
pub struct Resubmit {
    pub inp_file: String,
    pub pbs_file: String,
    pub job_id: String,
}

pub trait Submit<P>: SubQueue<P>
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    /// submit `filename` to the queue and return the jobid
    fn submit(&self, filename: &str) -> String {
        loop {
            match Command::new(self.submit_command()).arg(filename).output() {
                Ok(s) => {
                    if s.status.success() {
                        let raw = str::from_utf8(&s.stdout)
                            .unwrap()
                            .trim()
                            .to_string();
                        return raw
                            .split_whitespace()
                            .last()
                            .unwrap_or("")
                            .to_string();
                    }
                    eprintln!(
                        "failed to submit {filename} with `{}`",
                        String::from_utf8_lossy(&s.stderr)
                    );
                    std::thread::sleep(Duration::from_secs(1));
                }
                Err(e) => panic!("{e:?}"),
            };
        }
    }
}

/// a trait for all of the program-independent parts of a [Queue]
pub trait SubQueue<P>
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    /// the extension to append to submit scripts for this type of Queue
    const SCRIPT_EXT: &'static str;

    fn dir(&self) -> &str;

    fn submit_command(&self) -> &str;

    fn chunk_size(&self) -> usize;

    fn job_limit(&self) -> usize;

    fn sleep_int(&self) -> usize;

    /// the command to check the status of jobs in the queue
    fn stat_cmd(&self) -> String;

    /// return a HashSet of jobs found in the queue based on the output of
    /// `stat_cmd`
    fn status(&self) -> HashSet<String>;

    /// return `true` if all output files should be preserved
    fn no_del(&self) -> bool;
}

pub trait Queue<P>: SubQueue<P> + Submit<P>
where
    P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
{
    fn default_submit_script(&self) -> String;

    fn template(&self) -> &Option<String>;

    fn program_cmd(&self, filename: &str) -> String;

    fn write_submit_script(
        &self,
        infiles: impl IntoIterator<Item = String>,
        filename: &str,
    ) {
        use std::fmt::Write;
        let path = Path::new(filename);
        let basename = path.file_name().unwrap();
        let mut body = self
            .template()
            .clone()
            .unwrap_or_else(|| <Self as Queue<P>>::default_submit_script(self))
            .replace("{{.basename}}", basename.to_str().unwrap())
            .replace("{{.filename}}", filename);
        for f in infiles {
            writeln!(body, "{}", self.program_cmd(&f)).unwrap();
        }
        if std::fs::write(filename, body).is_err() {
            panic!("write_submit_script: failed to create {filename}");
        };
    }

    /// take a name of a Program input file with the extension attached, replace
    /// the extension (ext) with _redo.ext and write _redo.SCRIPT_EXT, then
    /// submit the redo script
    fn resubmit(&self, filename: &str) -> Resubmit {
        let path = Path::new(filename);
        let dir = path.parent().unwrap().to_str().unwrap();
        let base = path.file_stem().unwrap().to_str().unwrap();
        {
            let ext = path.extension().unwrap().to_str().unwrap();
            let inp_file = format!("{dir}/{base}_redo.{ext}");
            if let Err(e) = std::fs::copy(filename, &inp_file) {
                panic!("failed to copy {filename} to {inp_file} with `{e}`")
            }
        }
        // nothing but the copy needs the name with extension
        let inp_name = format!("{dir}/{base}_redo");
        let pbs_file = format!("{}/{}_redo.{}", dir, base, Self::SCRIPT_EXT);
        self.write_submit_script([inp_name.clone()], &pbs_file);
        let job_id = self.submit(&pbs_file);
        Resubmit {
            inp_file: inp_name,
            pbs_file,
            job_id,
        }
    }

    /// Build a chunk of jobs by writing the Program input file and the
    /// corresponding submission script and then submitting the script. returns
    /// the total durations spent writing input files, writing the submit
    /// script, and submitting the script
    fn build_chunk(
        &self,
        dir: &str,
        jobs: &mut [Job<P>],
        chunk_num: usize,
        proc: Procedure,
    ) -> (HashMap<String, usize>, Duration, Duration, Duration) {
        self.build_chunk_inner(dir, "main", chunk_num, jobs, proc)
    }

    fn build_chunk_inner(
        &self,
        dir: &str,
        base: &str,
        chunk_num: usize,
        jobs: &mut [Job<P>],
        proc: Procedure,
    ) -> (HashMap<String, usize>, Duration, Duration, Duration) {
        let mut input = Duration::default();
        let mut script = Duration::default();
        let mut submit = Duration::default();
        let queue_file =
            format!("{}/{base}{}.{}", dir, chunk_num, Self::SCRIPT_EXT);
        let jl = jobs.len();
        let mut slurm_jobs = HashMap::new();
        let filenames = jobs.iter_mut().map(|job| {
            time!(e, {
                job.program.write_input(proc);
            });
            input += e;
            job.pbs_file = queue_file.to_string();
            job.program.filename()
        });
        slurm_jobs.insert(queue_file.clone(), jl);
        time!(e, {
            self.write_submit_script(filenames, &queue_file);
        });
        script += e;
        // run jobs
        let job_id;
        time!(e, {
            job_id = self.submit(&queue_file);
        });
        submit += e;
        for job in jobs {
            job.job_id = job_id.clone();
        }
        (slurm_jobs, input, script, submit)
    }

    fn drain_err_case(
        &self,
        e: ProgramError,
        qstat: &mut HashSet<String>,
        slurm_jobs: &mut HashMap<String, usize>,
        job: &mut Job<P>,
    ) {
        let no_resub = LazyCell::new(|| std::env::var("SEMP_RESUB").is_ok());
        // just overwrite the existing job with the resubmitted
        // version
        if !qstat.contains(&job.job_id) {
            let time = job.modtime();
            if time > job.modtime {
                // file has been updated since we last looked at it, so need to
                // look again
                job.modtime = time;
                return;
            }
            eprintln!(
                "resubmitting {} (id={}) for {:?}",
                job.program.filename(),
                job.job_id,
                e
            );
            if *no_resub {
                eprintln!("resubmission disabled by SEMP_RESUB environment variable, exiting");
                std::process::exit(1);
            }
            let resub = format!(
                "{}.{}",
                job.program.filename(),
                job.program.extension()
            );
            let Resubmit {
                inp_file,
                pbs_file,
                job_id,
            } = self.resubmit(&resub);
            job.program.set_filename(&inp_file);
            job.pbs_file = pbs_file.clone();
            slurm_jobs.insert(pbs_file, 1);
            qstat.insert(job_id.clone());
            job.job_id = job_id;
        }
    }

    /// optimize is a copy of drain for optimizing jobs
    fn optimize(
        &self,
        dir: &str,
        jobs: Vec<Job<P>>,
        dst: &mut [Geom],
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        Opt.drain(dir, self, jobs, dst, Check::None)
    }

    /// resume draining from the checkpoint file in `checkpoint`
    fn resume(
        &self,
        dir: &str,
        checkpoint: &str,
        dst: &mut [f64],
        check: Check,
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        let jobs = Single::load_checkpoint(checkpoint, dst);
        eprintln!(
            "resuming from checkpoint in '{checkpoint}' with {} jobs remaining",
            jobs.len()
        );
        self.drain(dir, jobs, dst, check)
    }

    /// run the single-point energy calculations in `jobs`, storing the results
    /// in `dst`. if `check_int` > 0, write checkpoint files at that interval
    fn drain(
        &self,
        dir: &str,
        jobs: Vec<Job<P>>,
        dst: &mut [f64],
        check: Check,
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        Single.drain(dir, self, jobs, dst, check)
    }

    fn energize(
        &self,
        dir: &str,
        jobs: Vec<Job<P>>,
        dst: &mut [ProgramResult],
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        Both.drain(dir, self, jobs, dst, Check::None)
    }
}
