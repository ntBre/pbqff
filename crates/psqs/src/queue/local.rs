use serde::{Deserialize, Serialize};
use std::collections::HashSet;

use crate::program::dftbplus::DFTBPlus;
use crate::program::molpro::Molpro;
use crate::program::{mopac::Mopac, Program};
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Minimal implementation for testing MOPAC locally
#[derive(Debug)]
pub struct Local {
    pub dir: String,
    pub chunk_size: usize,
    pub template: Option<String>,
}

impl Default for Local {
    fn default() -> Self {
        Self {
            dir: ".".to_string(),
            chunk_size: 128,
            template: None,
        }
    }
}

impl Local {
    pub fn new(
        chunk_size: usize,
        _job_limit: usize,
        _sleep_int: usize,
        dir: &'static str,
        _no_del: bool,
        template: Option<String>,
    ) -> Self {
        Self {
            dir: dir.to_string(),
            chunk_size,
            template,
        }
    }
}

impl Submit<Molpro> for Local {}

impl Queue<Molpro> for Local {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOLPRO_CMD {filename}.inp")
    }

    fn default_submit_script(&self) -> String {
        String::new()
    }
}

impl Submit<Mopac> for Local {}

impl Queue<Mopac> for Local {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOPAC_CMD {filename}.mop")
    }

    fn default_submit_script(&self) -> String {
        "export MOPAC_CMD=/opt/mopac/mopac
export LD_LIBRARY_PATH=/opt/mopac/\n"
            .into()
    }
}

impl Submit<DFTBPlus> for Local {}

impl Queue<DFTBPlus> for Local {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("(cd {filename} && $DFTB_CMD > out)")
    }

    fn default_submit_script(&self) -> String {
        "DFTB_CMD=/opt/dftb+/dftb+\n".into()
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> SubQueue<P>
    for Local
{
    fn submit_command(&self) -> &str {
        "bash"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        1600
    }

    fn sleep_int(&self) -> usize {
        1
    }

    const SCRIPT_EXT: &'static str = "slurm";

    fn dir(&self) -> &str {
        &self.dir
    }

    fn stat_cmd(&self) -> String {
        todo!()
    }

    fn status(&self) -> HashSet<String> {
        for dir in ["opt", "pts", "freqs"] {
            let Ok(d) = std::fs::read_dir(dir) else {
                log::error!("{dir} not found for status");
                continue;
            };
            for f in d {
                eprintln!("contents of {:?}", f.as_ref().unwrap());
                eprintln!(
                    "{}",
                    std::fs::read_to_string(f.unwrap().path()).unwrap()
                );
                eprintln!("================");
            }
        }
        panic!("no status available for Local queue");
    }

    fn no_del(&self) -> bool {
        false
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::program::cfour::Cfour;

    use super::*;

    fn local() -> Local {
        Local {
            dir: String::new(),
            chunk_size: 0,
            template: None,
        }
    }

    macro_rules! make_tests {
        ($($name:ident, $queue:expr => $p:ty$(,)*)*) => {
            $(
            #[test]
            fn $name() {
                let tmp = tempfile::NamedTempFile::new().unwrap();
                <Local as Queue<$p>>::write_submit_script(
                    $queue,
                    ["opt0.inp", "opt1.inp", "opt2.inp", "opt3.inp"].map(|s| s.into()),
                    tmp.path().to_str().unwrap(),
                );
                let got = std::fs::read_to_string(tmp).unwrap();
                let got: Vec<&str> = got.lines().filter(|l|
                    !l.contains("/tmp")).collect();
                let got = got.join("\n");
                assert_snapshot!(got);
            }
            )*
        }
    }

    make_tests! {
        mopac_local, &local() =>  Mopac,
        molpro_local, &local() =>  Molpro,
        cfour_local, &local() => Cfour,
        dftb_local, &local() => DFTBPlus,
    }
}
