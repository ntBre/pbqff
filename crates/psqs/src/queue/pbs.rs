use std::path::Path;
use std::time::Duration;
use std::{collections::HashSet, process::Command};

use serde::{Deserialize, Serialize};

use crate::program::dftbplus::DFTBPlus;
use crate::program::molpro::Molpro;
use crate::program::mopac::Mopac;
use crate::program::Program;
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Pbs is a type for holding the information for submitting a pbs job.
/// `filename` is the name of the Pbs submission script
#[derive(Debug)]
pub struct Pbs {
    pub chunk_size: usize,
    pub job_limit: usize,
    pub sleep_int: usize,
    pub dir: &'static str,
    pub no_del: bool,
    pub template: Option<String>,
}

impl Pbs {
    pub fn new(
        chunk_size: usize,
        job_limit: usize,
        sleep_int: usize,
        dir: &'static str,
        no_del: bool,
        template: Option<String>,
    ) -> Self {
        Self {
            chunk_size,
            job_limit,
            sleep_int,
            dir,
            no_del,
            template,
        }
    }
}

impl Submit<Mopac> for Pbs
where
    Mopac: Serialize + for<'a> Deserialize<'a>,
{
    /// submit `filename` to the queue and return the jobid
    fn submit(&self, filename: &str) -> String {
        let mut cmd =
            Command::new(<Self as SubQueue<Mopac>>::submit_command(self));
        let cmd = cmd.arg("-f").arg(filename);
        submit_inner(cmd, self.sleep_int).unwrap()
    }
}

// Molpro 2022 submit script requires submission from the current directory, so
// we have to override the default impl
impl Submit<Molpro> for Pbs
where
    Molpro: Serialize + for<'a> Deserialize<'a>,
{
    fn submit(&self, filename: &str) -> String {
        let path = Path::new(filename);
        let dir = path.parent().unwrap();
        let base = path.file_name().unwrap();
        let mut cmd =
            Command::new(<Self as SubQueue<Molpro>>::submit_command(self));
        let cmd = cmd.arg(base).current_dir(dir);
        submit_inner(cmd, self.sleep_int).unwrap()
    }
}

/// helper function to consolidate error handling between the two submit
/// implementations
fn submit_inner(
    cmd: &mut Command,
    sleep_int: usize,
) -> std::io::Result<String> {
    let mut retries = 5;
    loop {
        match cmd.output() {
            Ok(s) => {
                if !s.status.success() {
                    if retries > 0 {
                        eprintln!(
                            "qsub failed with output: {s:#?}, \
				   retrying {retries} more times"
                        );
                        retries -= 1;
                        std::thread::sleep(Duration::from_secs(
                            sleep_int as u64,
                        ));
                        continue;
                    }
                    panic!("qsub failed with output: {s:#?}");
                }
                let raw =
                    std::str::from_utf8(&s.stdout).unwrap().trim().to_string();
                return Ok(raw
                    .split_whitespace()
                    .last()
                    .unwrap_or("no jobid")
                    .to_string());
            }
            Err(e) => return Err(e),
        }
    }
}

impl Queue<Molpro> for Pbs
where
    Molpro: Serialize + for<'a> Deserialize<'a>,
{
    fn template(&self) -> &Option<String> {
        &self.template
    }

    /// This must be consistent with the Submit<Molpro> implementation, which
    /// currently changes to the parent directory of the PBS script before
    /// submitting. This also assumes, then, that the PBS script is in the same
    /// directory as the input files, but I think that's a safe assumption.
    fn program_cmd(&self, filename: &str) -> String {
        let basename = Path::new(&filename).file_name().unwrap();
        format!("$MOLPRO_CMD {basename:?}.inp")
    }

    fn default_submit_script(&self) -> String {
        "#!/bin/sh
#PBS -N {{.basename}}
#PBS -S /bin/bash
#PBS -j oe
#PBS -o {{.basename}}.out
#PBS -W umask=022
#PBS -l walltime=1000:00:00
#PBS -l ncpus=1
#PBS -l mem=8gb
#PBS -q workq

module load openpbs molpro

export WORKDIR=$PBS_O_WORKDIR
export TMPDIR=/tmp/$USER/$PBS_JOBID
cd $WORKDIR
mkdir -p $TMPDIR
trap 'rm -rf $TMPDIR' EXIT

export MOLPRO_CMD=\"molpro -t $NCPUS --no-xml-output\"
"
        .to_owned()
    }
}

impl Queue<Mopac> for Pbs {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOPAC_CMD {filename}.mop")
    }

    fn default_submit_script(&self) -> String {
        "#!/bin/sh
#PBS -N {{.basename}}
#PBS -S /bin/bash
#PBS -j oe
#PBS -o {{.filename}}.out
#PBS -W umask=022
#PBS -l walltime=1000:00:00
#PBS -l ncpus=1
#PBS -l mem=1gb
#PBS -q workq

module load openpbs

export WORKDIR=$PBS_O_WORKDIR
cd $WORKDIR

export LD_LIBRARY_PATH=/ddnlus/r2518/Packages/mopac/build
export MOPAC_CMD=/ddnlus/r2518/Packages/mopac/build/mopac
"
        .to_owned()
    }
}

impl Queue<DFTBPlus> for Pbs {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("(cd {filename} && $DFTB_CMD > out)")
    }

    fn default_submit_script(&self) -> String {
        "#!/bin/sh
#PBS -N {{.basename}}
#PBS -S /bin/bash
#PBS -j oe
#PBS -o {{.filename}}.out
#PBS -W umask=022
#PBS -l walltime=1000:00:00
#PBS -l ncpus=1
#PBS -l mem=8gb
#PBS -q workq

module load openpbs

export WORKDIR=$PBS_O_WORKDIR
cd $WORKDIR

export DFTB_CMD=/ddnlus/r2518/.conda/envs/dftb/bin/dftb+
"
        .to_owned()
    }
}

impl Submit<DFTBPlus> for Pbs {
    fn submit(&self, filename: &str) -> String {
        let mut cmd =
            Command::new(<Self as SubQueue<DFTBPlus>>::submit_command(self));
        let cmd = cmd.arg("-f").arg(filename);
        submit_inner(cmd, self.sleep_int).unwrap()
    }
}

impl<P> SubQueue<P> for Pbs
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    fn submit_command(&self) -> &str {
        "qsub"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        self.job_limit
    }

    fn sleep_int(&self) -> usize {
        self.sleep_int
    }

    const SCRIPT_EXT: &'static str = "pbs";

    fn dir(&self) -> &str {
        self.dir
    }

    /// run `qstat -u $USER`. form of the output is:
    ///
    /// maple:
    ///                                                     Req'd  Req'd   Elap
    /// Job ID  Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
    /// ------- -------- -------- ---------- ------ --- --- ------ ----- - -----
    /// 819446  user     queue    C6HNpts      5085   1   1    8gb 26784 R 00:00
    fn stat_cmd(&self) -> String {
        let user = std::env::var("USER").expect("couldn't find $USER env var");
        let status = match Command::new("qstat").args(["-u", &user]).output() {
            Ok(status) => status,
            Err(e) => panic!("failed to run `qstat -u {user}` with {e}"),
        };
        String::from_utf8(status.stdout).expect("failed to parse qstat output")
    }

    fn status(&self) -> HashSet<String> {
        let mut ret = HashSet::new();
        let lines = <Pbs as SubQueue<P>>::stat_cmd(self);
        // skip to end of header
        let lines = lines.lines().skip_while(|l| !l.contains("-----------"));
        for line in lines {
            let fields: Vec<_> = line.split_whitespace().collect();
            assert!(fields.len() == 11);
            ret.insert(fields[0].to_string());
        }
        ret
    }

    fn no_del(&self) -> bool {
        self.no_del
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::program::cfour::Cfour;

    use super::*;

    fn pbs() -> Pbs {
        Pbs {
            chunk_size: 1,
            job_limit: 1,
            sleep_int: 1,
            dir: "/tmp",
            no_del: false,
            template: None,
        }
    }

    macro_rules! make_tests {
        ($($name:ident, $queue:expr => $p:ty$(,)*)*) => {
            $(
            #[test]
            fn $name() {
                let tmp = tempfile::NamedTempFile::new().unwrap();
                <Pbs as Queue<$p>>::write_submit_script(
                    $queue,
                    ["pts/opt0.inp", "pts/opt1.inp", "pts/opt2.inp", "pts/opt3.inp"]
                    .map(|s| s.into()),
                    tmp.path().to_str().unwrap(),
                );
                let got = std::fs::read_to_string(tmp).unwrap();
                let got: Vec<&str> = got.lines().filter(|l|
                    !(l.starts_with("#PBS -N")
                        || l.starts_with("#PBS -o"))).collect();
                let got = got.join("\n");
                assert_snapshot!(got);
            }
            )*
        }
    }

    make_tests! {
        mopac_pbs, &pbs() =>  Mopac,
        molpro_pbs, &pbs() =>  Molpro,
        cfour_pbs, &pbs() => Cfour,
        dftb_pbs, &pbs() => DFTBPlus,
    }
}
