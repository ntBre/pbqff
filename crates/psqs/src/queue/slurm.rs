use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use crate::program::mopac::Mopac;
use crate::program::Program;
use crate::program::{dftbplus::DFTBPlus, molpro::Molpro};
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Slurm is a type for holding the information for submitting a slurm job.
/// `filename` is the name of the Slurm submission script
#[derive(Debug)]
pub struct Slurm {
    chunk_size: usize,
    job_limit: usize,
    sleep_int: usize,
    dir: &'static str,
    no_del: bool,
    pub(crate) template: Option<String>,
}

impl Slurm {
    pub fn new(
        chunk_size: usize,
        job_limit: usize,
        sleep_int: usize,
        dir: &'static str,
        no_del: bool,
        template: Option<String>,
    ) -> Self {
        Self {
            chunk_size,
            job_limit,
            sleep_int,
            dir,
            no_del,
            template,
        }
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> Submit<P>
    for Slurm
{
}

impl Queue<Molpro> for Slurm {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOLPRO_CMD {filename}.inp")
    }

    fn default_submit_script(&self) -> String {
        include_str!("../../templates/slurm/molpro").to_owned()
    }
}

impl Queue<Mopac> for Slurm {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOPAC_CMD {filename}.mop")
    }

    fn default_submit_script(&self) -> String {
        include_str!("../../templates/slurm/mopac").to_owned()
    }
}

impl Queue<DFTBPlus> for Slurm {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("(cd {filename} && $DFTB_CMD > out)")
    }

    fn default_submit_script(&self) -> String {
        String::new()
    }
}

impl<P> SubQueue<P> for Slurm
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    fn submit_command(&self) -> &str {
        "sbatch"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        self.job_limit
    }

    fn sleep_int(&self) -> usize {
        self.sleep_int
    }

    const SCRIPT_EXT: &'static str = "slurm";

    fn dir(&self) -> &str {
        self.dir
    }

    /// run `squeue -u $USER`. form of the output is:
    ///
    ///    JOBID PARTITION   NAME     USER ST        TIME  NODES NODELIST(REASON)
    /// 30627992   compute  c3oh-   mdavis  R 46-17:12:23      1 node2
    fn stat_cmd(&self) -> String {
        let user = std::env::var("USER").expect("couldn't find $USER env var");
        let status = match std::process::Command::new("squeue")
            .args(["-u", &user])
            .output()
        {
            Ok(status) => status,
            Err(e) => panic!("failed to run squeue with {e}"),
        };
        String::from_utf8(status.stdout)
            .expect("failed to convert squeue output to String")
    }

    fn status(&self) -> HashSet<String> {
        let mut ret = HashSet::new();
        // wut?
        let lines = <Slurm as SubQueue<P>>::stat_cmd(self);
        let lines = lines.lines();
        for line in lines {
            if !line.contains("JOBID") {
                let fields: Vec<_> = line.split_whitespace().collect();
                assert!(fields.len() == 8);
                // exclude completing jobs to combat stuck completing bug
                if fields[4] != "CG" {
                    ret.insert(fields[0].to_string());
                }
            }
        }
        ret
    }

    fn no_del(&self) -> bool {
        self.no_del
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::program::cfour::Cfour;

    use super::*;

    fn slurm() -> Slurm {
        Slurm {
            chunk_size: 1,
            job_limit: 1,
            sleep_int: 1,
            dir: "/tmp",
            no_del: false,
            template: None,
        }
    }

    macro_rules! make_tests {
        ($($name:ident, $queue:expr => $p:ty$(,)*)*) => {
            $(
            #[test]
            fn $name() {
                let tmp = tempfile::NamedTempFile::new().unwrap();
                <Slurm as Queue<$p>>::write_submit_script(
                    $queue,
                    ["opt0.inp", "opt1.inp", "opt2.inp", "opt3.inp"].map(|s| s.into()),
                    tmp.path().to_str().unwrap(),
                );
                let got = std::fs::read_to_string(tmp).unwrap();
                let got: Vec<&str> = got.lines().filter(|l|
                    !(l.starts_with("#SBATCH --job-name")
                        || l.starts_with("#SBATCH -o"))).collect();
                let got = got.join("\n");
                assert_snapshot!(got);
            }
            )*
        }
    }

    make_tests! {
        mopac_slurm, &slurm() =>  Mopac,
        molpro_slurm, &slurm() => Molpro,
        cfour_slurm, &slurm() => Cfour,
        dftb_slurm, &slurm() => DFTBPlus,
    }
}
