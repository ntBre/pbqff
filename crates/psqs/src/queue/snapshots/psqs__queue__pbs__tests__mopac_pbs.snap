---
source: src/queue/pbs.rs
expression: got
---
#!/bin/sh
#PBS -S /bin/bash
#PBS -j oe
#PBS -W umask=022
#PBS -l walltime=1000:00:00
#PBS -l ncpus=1
#PBS -l mem=1gb
#PBS -q workq

module load openpbs

export WORKDIR=$PBS_O_WORKDIR
cd $WORKDIR

export LD_LIBRARY_PATH=/ddnlus/r2518/Packages/mopac/build
export MOPAC_CMD=/ddnlus/r2518/Packages/mopac/build/mopac
$MOPAC_CMD pts/opt0.inp.mop
$MOPAC_CMD pts/opt1.inp.mop
$MOPAC_CMD pts/opt2.inp.mop
$MOPAC_CMD pts/opt3.inp.mop
